pkg/driver/node/credentialprovider/provider_pod.go

```diff
@@ -48,56 +55,68 @@ func (c *Provider) provideFromPod(ctx context.Context, provideCtx ProvideContext
-	if podID == "" {
-		return nil, status.Error(codes.InvalidArgument, "Missing Pod info. Please make sure to enable `podInfoOnMountCompat`, see "+podLevelCredentialsDocsPage)
+		// TODO: These were needed with `systemd` but probably won't be necessary with containerization.
+		envprovider.EnvMountpointCacheKey:    cacheKey,
+		envprovider.EnvConfigFile:            filepath.Join(provideCtx.EnvPath, "disable-config"),
+		envprovider.EnvSharedCredentialsFile: filepath.Join(provideCtx.EnvPath, "disable-credentials"),
 	}
 
-	tokenName := podLevelServiceAccountTokenName(podID, volumeID)
+	// 3. Create IRSA and EKS Pod Identity environments
+	irsaCredentialsEnvironment, irsaCredentialsEnvironmentError := c.createIRSACredentialsEnvironment(ctx, provideCtx)
+	eksPodIdentityCredentialsEnvironment, eksPodIdentityCredentialsEnvironmentError := c.createEKSPodIdentityCredentialsEnvironment(provideCtx)
 
-	err = renameio.WriteFile(filepath.Join(provideCtx.WritePath, tokenName), []byte(stsToken.Token), CredentialFilePerm)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to write service account token: %v", err)
+	if irsaCredentialsEnvironmentError != nil && eksPodIdentityCredentialsEnvironmentError != nil { // TODO: Consider chaining this to the If statements below
+		klog.Error("IRSA and EKS Pod Identity failed")                                                                                                                    // TODO: Improve error message
+		return nil, status.Errorf(codes.Internal, "IRSA and EKS Pod Identity failed: %v, %v", irsaCredentialsEnvironmentError, eksPodIdentityCredentialsEnvironmentError) // TODO: Improve error message
 	}
 
-	podNamespace := provideCtx.PodNamespace
-	podServiceAccount := provideCtx.ServiceAccountName
-	cacheKey := podNamespace + "/" + podServiceAccount
+	// 4. Include only the appropriate environment (IRSA or EKS Pod Identity) in the environment to be returned and copy only the appropriate token to WritePath
+	// (if both methods are configured, IRSA takes precedence)
+	if irsaCredentialsEnvironmentError == nil {
+		klog.V(4).Infof("Providing credentials from pod with STS Web Identity provider (IRSA)")
 
-	return envprovider.Environment{
-		envprovider.EnvRoleARN:              roleARN,
-		envprovider.EnvWebIdentityTokenFile: filepath.Join(provideCtx.EnvPath, tokenName),
+		// Copy STS Token file to WritePath
+		tokenName := podLevelServiceAccountTokenName(provideCtx.PodID, provideCtx.VolumeID)
+		err := renameio.WriteFile(filepath.Join(provideCtx.WritePath, tokenName), []byte(stsToken.Token), CredentialFilePerm)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to write service account STS token: %v", err)
+		}
 
-		envprovider.EnvRegion:        region,
-		envprovider.EnvDefaultRegion: defaultRegion,
+		env.Merge(irsaCredentialsEnvironment)
+	} else {
+		klog.Error("Error providing credentials from pod with STS Web Identity provider (IRSA)")
```

Comment:
Don't think we need to log this as error log. Info level should be fine since it's expected if customer only configured Pod Identity credentials.